experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+        attr_vals += convert_getattrs_to_inputs(k4)
+        print_value('attr_vals_after_functionalize', attr_vals)
+
+        print_graph('after functionalize before DCE', k4)
+
+        # 7. eliminate dead code but make sure we're still keeping the final copy_
+        torch.fx.node._side_effectful_functions.add(torch.ops.aten.copy_.default)
+        k4.graph.eliminate_dead_code()
+        k4.graph.lint()
+        k4.recompile()
+
+        print_graph('after functionalize after DCE', k4)
+
+        return k4, attr_vals, unused_inputs_idx
+
+    return call
+
+DEFAULT_SCHEMA = 'default' 
+
+
+def track_state_dict_materialization(map: Dict[FakeTensor, Tensor], fake_state_dict, real_state_dict):
+    for k, v in fake_state_dict.items():
+        real_k = map[k] if isinstance(k, Tensor) else k
+        if isinstance(v, Tensor):
+            map[v] = real_state_dict[real_k]
+        elif isinstance(v, dict):
+            track_state_dict_materialization(map, v, real_state_dict[real_k])
+
+
+def warmup_optimizer(optim):
+    # unfortunatelly, in order to initialize the optimizer states
+    # we need to call optimizer.step(). In order to do that, we will initialize
+    # parameter gradients with zero tensors just so we can run a step.
+    for group in optim.param_groups:
+        for param in group['params']:
+            param.grad = torch.zeros_like(param)
+    # run an optimizer step just to initialize the optimizer states
+    # multiple issues here:
+    # 1. this will make the optimizer think we're at the 2nd iteration
+    # 2. doesn't support optimizers that require callables passed in
+    optim.step()
+    # now we can delete those zero gradients
+    optim.zero_grad(set_to_none=True)
+
+
+def materialize_optimizer(optim, tracking_map):
+    # reset this optimizer state so that we can properly materialize
+    # the optimizer state next.
+    from collections import defaultdict
+    saved_fake_state = optim.state
+    optim.state = defaultdict(dict)
+
+    # replace parameters in the param groups with the materialized
+    # parameters taken from the tracking materializer
+    # this will throw if any of the parameters is not found in one of the models
+    for group in optim.param_groups:
+        group['params'] = [tracking_map[param] for param in group['params']]
+
+    # this warms up the optimizer with real state dict now
+    warmup_optimizer(optim)
+
+    # we will then map the fake optimizer states to the real ones
+    track_state_dict_materialization(
+        tracking_map,
+        saved_fake_state,
+        optim.state,
+    )
+
+
+def _materialize_tensor(t: torch.Tensor, param=None):
+    if True: # if isinstance(t, FakeTensor):
+        return torch.zeros(t.shape, device=t.device)
+
+	
+T = TypeVar('T')
+
+
+def _materialize_dtensor(mesh, default, tensor, param=None):
+    if isinstance(tensor, DTensor):
+        return tensor
+    else:
+        # do we have a way to make this into a DTensor?
```

Comment:
do you mean full tensor -> DTensor? maybe use the `distribute_tensor` API?

Reply:
Right. Basically the logic here is as follows:
1.1) If we have a tag on this tensor to specify its sharding, _or_ if we have a default sharding strategy then we produce a DTensor.
1.2) Otherwise produce a Tensor (not DTensor)

As for what to fill the materialized tensor with:
  2.1) If this is a Fake tensor, then just make a Tensor [1.2] (or DTensor [1.1]) containing zeros. (User could further initialize this with torch.nn.init functions)
  2.2) If this was a real already a tensor and we are producing a DTensor [1.1], then use distribute_tensor API. Otherwise, just keep the original tensor as is.